_posts/2022-05-20-site-settings.md
---
layout: post
title: Configuring User-Configurable Settings
tags:
 - documentation
---

The Hematite theme comes with a user-configurable settings page. Settings are stored using [`localStorage`](https://developer.mozilla.org/en-US/docs/Web/API/Window/localStorage).

<details markdown=1><summary>The settings GUI looks like this</summary>
```md
<!--
    Include the settings GUI in a page with the following Liquid include:
-->
{% raw %}{% include settings.html %}{% endraw %}
```
{% include settings.html %}
</details>

## Disabling them
The user-configurable settings page (enabled by default) can be disabled by adding the following to `_config.yml`:
```yaml
hematite:
  # ...

  sidebar:
    # ...
    show_settings_btn: false
```

You can see what settings are available [on this site's settings page]({{ "assets/html/settings" | relative_url }}).

## Adding custom settings
{% raw %}
If you have your own settings, create your own `assets/html/settings.html` to override the default settings page. You can then `{% include settings.html %}` to include the settings widgets that come with the Hematite theme.

The default settings page, for example, looks similar to this:
```html
---
layout: default
title: Settings
noindex: true
---

<!-- in assets/html/settings.html -->

<h1 id="settings_header">page_settings_header</h1>
<p id="settings_description">page_settings_description</p>

{% include settings.html %}
```

## Setting settings' default values
Change the default values of many of the settings in `_config.yml`!
```yaml
hematite:
    # ... Other settings ...

    user_config:
        # The font associated with the "small"
        # font size option, in pt.
        font_size_small: 9
        font_size_medium: 13
        font_size_large: 16

        # Whether to show/hide the sticky header by default
        minimize_header: false
```

Currently, the default font family must be changed with CSS. For example, if your site's `assets/style.scss` looks like this:
```scss
---
title: false
styles: true
---

@import "hematite";

// In dark mode, invert the brightness of images
@include auto-invert-images;
```
The default font families can then be customized by adding this to the end of `style.scss`:
```scss
:root {
    --serif-font-family: 'Times New Roman', 'Font of your choice', serif;
    --sans-font-family: 'URW Gothic, Book', 'Font of your choice', sans-serif;
    --main-font-family: var(--serif-font-family);
    // or
    // --main-font-family: var(--sans-font-family);
    // to default to a sans font.
}
```

{% endraw %}


assets/js/Settings.mjs
---
---
import AsyncUtil from "./AsyncUtil.mjs";
import AnimationUtil from "./AnimationUtil.mjs";

const THEME_TRANSITION_TIME = 500; // ms

class Settings {
    THEME_AUTO = 'auto';
    THEME_DARK = 'dark';
    THEME_LIGHT = 'light';

    FONT_DEFAULT = 'default';
    FONT_SANS = 'sans';
    FONT_SERIF = 'serif';

    FONT_SIZE_SMALL = 'small';
    FONT_SIZE_MEDIUM = 'medium';
    FONT_SIZE_LARGE = 'large';
    FONT_SIZE_DEFAULT = 'default';

    FONT_SIZE_KEY_ = 'hematite-setting-font-size';
    FONT_FAMILY_KEY_ = 'hematite-setting-font-family';
    THEME_KEY_ = 'hematite-setting-theme';
    HEADER_MINIMIZED_KEY_ = 'hematite-setting-minimize-header';

    constructor() {
    }

    getSetting_(key) {
        return localStorage.getItem(key);
    }

    setSetting_(key, val) {
        localStorage.setItem(key, val);
        this.applySettings();
    }

    getTheme() {
        return this.getSetting_(this.THEME_KEY_) ?? this.THEME_AUTO;
    }

    getFontFamily() {
        return this.getSetting_(this.FONT_FAMILY_KEY_) ?? this.FONT_DEFAULT;
    }

    getFontSize() {
        return this.getSetting_(this.FONT_SIZE_KEY_) ?? this.FONT_SIZE_DEFAULT;
    }

    getHeaderMinimized() {
        let minimizedStr = this.getSetting_(this.HEADER_MINIMIZED_KEY_)
            ?? {{ site.hematite.user_config.minimize_header | default: site.hematite.minimize_header | default: "false" | jsonify }};
        return minimizedStr == "true";
    }

    setFontSize(sizeOption) {
        this.setSetting_(this.FONT_SIZE_KEY_, sizeOption);
    }

    setFontFamily(familyOption) {
        this.setSetting_(this.FONT_FAMILY_KEY_, familyOption);
    }

    setTheme(themeOption) {
        this.setSetting_(this.THEME_KEY_, themeOption);
    }

    setHeaderMinimized(minimize) {
        this.setSetting_(this.HEADER_MINIMIZED_KEY_, `${minimize}`);
    }

    getFontSizePt_() {
        let fontSizeOption = this.getFontSize();

        if (fontSizeOption == this.FONT_SIZE_SMALL) {
            return {{ site.hematite.user_config.font_size_small | default: "9" }};
        }

        if (fontSizeOption == this.FONT_SIZE_MEDIUM) {
            return {{ site.hematite.user_config.font_size_medium | default: "13" }};
        }

        if (fontSizeOption == this.FONT_SIZE_LARGE) {
            return {{ site.hematite.user_config.font_size_large | default: "16" }};
        }
    }

    usingNonDefaultFontFamily_() {
        return this.getFontFamily() != this.FONT_DEFAULT;
    }

    usingNonDefaultFontSize_() {
        return this.getFontSize() != this.FONT_SIZE_DEFAULT;
    }

    forcingDarkTheme_() {
        return this.getTheme() == this.THEME_DARK;
    }

    forcingLightTheme_() {
        return this.getTheme() == this.THEME_LIGHT;
    }

    async applySettings() {
        document.documentElement.classList.add("changingTheme");

        // Clean up previous changes. We might be re-applying styles.
        document.documentElement.classList.remove("lightTheme");
        document.documentElement.classList.remove("darkTheme");

        if (!this.style_) {
            this.style_ = document.createElement("style");
        } else {
            this.style_.remove();
        }


        let styleHTML =
        `
            :root.changingTheme * {
                transition: ${THEME_TRANSITION_TIME}ms ease all;
            }
        `;

        // Font family
        if (this.usingNonDefaultFontFamily_()) {
            styleHTML +=
            `
                :root {
                    --main-font-family: var(--${this.getFontFamily()}-font-family);
                }
            `;
        }

        // Font size
        if (this.usingNonDefaultFontSize_()) {
            styleHTML +=
            `
                :root {
                    --main-font-size: ${this.getFontSizePt_()}pt;
                }
            `;
        }

        this.style_.innerHTML = styleHTML;
        document.documentElement.appendChild(this.style_);

        // Theme
        if (this.forcingDarkTheme_()) {
            document.documentElement.classList.add("darkTheme");
        }
        else if (this.forcingLightTheme_()) {
            document.documentElement.classList.add("lightTheme");
        }

        if (this.getHeaderMinimized()) {
            // Run roughly in parallel
            (async () => {
                let header = document.querySelector("body > header");
                if (header) {
                    await AnimationUtil.collapseOutVert(header, THEME_TRANSITION_TIME / 2);
                }

                document.documentElement.classList.add("minimizedNavHeader");

                if (header) {
                    // Re-show the header
                    // Rely on a CSS animation for animating the header's return.
                    await AnimationUtil.expandInVert(header, 0);
                }
            })();
        }
        else {
            document.documentElement.classList.remove("minimizedNavHeader");
        }

        await AsyncUtil.waitMillis(THEME_TRANSITION_TIME);
        document.documentElement.classList.remove("changingTheme");
    }
}

export default (new Settings());
